File: tauri/src-tauri/src/llm/lmstudio.rs
use crate::llm::{LlmProvider, Message, LlmEvent};
use async_trait::async_trait;
use futures_util::{Stream, StreamExt};
use reqwest::Client;
use serde_json::Value;
use std::pin::Pin;
use std::time::Duration;

pub struct LMStudio {
    base_url: String,
    client: Client,
}

impl LMStudio {
    pub fn new(base_url: Option<String>) -> Self {
        Self {
            base_url: base_url.unwrap_or_else(|| "http://localhost:1234/v1".to_string()),
            client: Client::new(),
        }
    }
}

#[async_trait]
impl LlmProvider for LMStudio {
    async fn stream_chat(
        &self,
        model: &str,
        messages: Vec<Message>,
    ) -> Result<Pin<Box<dyn Stream<Item = LlmEvent> + Send>>, String> {
        let response = self.client
            .post(format!("{}/chat/completions", self.base_url))
            .header("Content-Type", "application/json")
            .json(&serde_json::json!({
                "model": model,
                "messages": messages,
                "stream": true
            }))
            .send()
            .await
            .map_err(|e| format!("LM Studio Request failed: {}. Is LM Studio running on port 1234?", e))?;

        let stream = response.bytes_stream().map(|result| {
            match result {
                Ok(bytes) => {
                    let chunk = String::from_utf8_lossy(&bytes);
                    let mut events = Vec::new();
                    
                    for line in chunk.lines() {
                        let trimmed = line.trim();
                        if trimmed.is_empty() || !trimmed.starts_with("data: ") {
                            continue;
                        }
                        
                        let data = &trimmed[6..];
                        if data == "[DONE]" {
                            continue;
                        }
                        
                        if let Ok(json) = serde_json::from_str::<Value>(data) {
                            if let Some(choices) = json.get("choices").and_then(|c| c.as_array()) {
                                if let Some(choice) = choices.get(0) {
                                    if let Some(delta) = choice.get("delta") {
                                        if let Some(content) = delta.get("content").and_then(|c| c.as_str()) {
                                            events.push(LlmEvent::Token { token: content.to_string() });
                                        }
                                    }
                                }
                            }
                        }
                    }
                    events
                }
                Err(e) => vec![LlmEvent::Error { message: e.to_string() }],
            }
        })
        .flat_map(futures_util::stream::iter);

        Ok(Box::pin(stream))
    }

    async fn fetch_models(&self) -> Result<Vec<String>, String> {
        let response = self.client
            .get(format!("{}/models", self.base_url))
            .timeout(Duration::from_millis(2000))
            .send()
            .await
            .map_err(|e| e.to_string())?;

        let json: Value = response.json().await.map_err(|e| e.to_string())?;
        let mut models = Vec::new();

        if let Some(data) = json.get("data").and_then(|d| d.as_array()) {
            for model in data {
                if let Some(id) = model.get("id").and_then(|i| i.as_str()) {
                    models.push(id.to_string());
                }
            }
        }

        Ok(models)
    }
}

File: electron/electron/framework/llm/lmstudio.ts
import type { LlmProvider, Message, LlmEvent } from './types'

class LMStudio implements LlmProvider {
    private baseUrl: string

    constructor(baseUrl?: string) {
        this.baseUrl = baseUrl || 'http://localhost:1234/v1'
    }

    async *streamChat(model: string, messages: Message[]): AsyncIterable<LlmEvent> {
        const response = await fetch(`${this.baseUrl}/chat/completions`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ model, messages, stream: true }),
        })

        if (!response.ok) {
            yield { type: 'error', message: `LM Studio HTTP ${response.status}. Is LM Studio running on port 1234?` }
            return
        }

        const reader = response.body?.getReader()
        if (!reader) return

        const decoder = new TextDecoder()
        let buffer = ''

        while (true) {
            const { done, value } = await reader.read()
            if (done) break

            buffer += decoder.decode(value, { stream: true })
            const lines = buffer.split('\n')
            buffer = lines.pop() || ''

            for (const line of lines) {
                const trimmed = line.trim()
                if (!trimmed || !trimmed.startsWith('data: ')) continue

                const data = trimmed.slice(6)
                if (data === '[DONE]') continue

                try {
                    const json = JSON.parse(data)
                    const delta = json.choices?.[0]?.delta
                    if (delta?.content) {
                        yield { type: 'token', token: delta.content }
                    }
                } catch {
                    // Skip malformed chunks
                }
            }
        }
    }

    async fetchModels(): Promise<string[]> {
        const base = this.baseUrl.replace(/\/v1$/, '')
        const models: string[] = []

        try {
            const res = await fetch(`${base}/api/v1/models`, {
                signal: AbortSignal.timeout(3000),
            })
            const json: any = await res.json()

            if (Array.isArray(json.models)) {
                for (const m of json.models) {
                    if (m.type !== 'llm') continue
                    if (m.key) models.push(m.key)
                }
            }
        } catch {
            // Native API not available
        }

        if (models.length === 0) {
            try {
                const res = await fetch(`${this.baseUrl}/models`, {
                    signal: AbortSignal.timeout(2000),
                })
                const json: any = await res.json()

                if (Array.isArray(json.data)) {
                    for (const m of json.data) {
                        if (m.id) models.push(m.id)
                    }
                }
            } catch {
                // LM Studio not running
            }
        }

        return models
    }
}

export default LMStudio
